peer/record.go

```diff
@@ -0,0 +1,221 @@
+// "consume" the containing envelope, which will return both the
+// routing.Envelope and the inner Record. The Record must be cast to
+// a PeerRecord pointer before use:
+//
+//     envelope, untypedRecord, err := ConsumeEnvelope(envelopeBytes, PeerRecordEnvelopeDomain)
+//     if err != nil {
+//       handleError(err)
+//       return
+//     }
+//     peerRec := untypedRecord.(*PeerRecord)
+//
+type PeerRecord struct {
+	// PeerID is the ID of the peer this record pertains to.
+	PeerID ID
+
+	// Addrs contains the public addresses of the peer this record pertains to.
+	Addrs []ma.Multiaddr
+
+	// Seq is a monotonically-increasing sequence counter that's used to order
+	// PeerRecords in time. The interval between Seq values is unspecified,
+	// but newer PeerRecords MUST have a greater Seq value than older records
+	// for the same peer.
+	Seq uint64
+}
+
+// NewPeerRecord returns a PeerRecord with a timestamp-based sequence number.
+// The returned record is otherwise empty and should be populated by the caller.
+func NewPeerRecord() *PeerRecord {
+	return &PeerRecord{Seq: TimestampSeq()}
+}
+
+// PeerRecordFromAddrInfo creates a PeerRecord from an AddrInfo struct.
+// The returned record will have a timestamp-based sequence number.
+func PeerRecordFromAddrInfo(info AddrInfo) *PeerRecord {
+	rec := NewPeerRecord()
+	rec.PeerID = info.ID
+	rec.Addrs = info.Addrs
+	return rec
+}
+
+// TimestampSeq is a helper to generate a timestamp-based sequence number for a PeerRecord.
+func TimestampSeq() uint64 {
+	return uint64(time.Now().UnixNano())
+}
+
+// UnmarshalRecord parses a PeerRecord from a byte slice.
+// This method is called automatically when consuming a record.Envelope
+// whose PayloadType indicates that it contains a PeerRecord.
+// It is generally not necessary or recommended to call this method directly.
+func (r *PeerRecord) UnmarshalRecord(bytes []byte) error {
+	if r == nil {
+		return fmt.Errorf("cannot unmarshal PeerRecord to nil receiver")
+	}
+
+	var msg pb.PeerRecord
+	err := proto.Unmarshal(bytes, &msg)
+	if err != nil {
+		return err
+	}
+	var id ID
+	err = id.UnmarshalBinary(msg.PeerId)
+	if err != nil {
+		return err
+	}
+	r.PeerID = id
+	r.Addrs = addrsFromProtobuf(msg.Addresses)
+	r.Seq = msg.Seq
+	return nil
+}
+
+// MarshalRecord serializes a PeerRecord to a byte slice.
+// This method is called automatically when constructing a routing.Envelope
+// using MakeEnvelopeWithRecord or PeerRecord.Sign.
+func (r *PeerRecord) MarshalRecord() ([]byte, error) {
+	idBytes, err := r.PeerID.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	msg := pb.PeerRecord{
+		PeerId:    idBytes,
+		Addresses: addrsToProtobuf(r.Addrs),
+		Seq:       r.Seq,
+	}
+	return proto.Marshal(&msg)
+}
+
+// Sign wraps the PeerRecord in a routing.Envelope, signed with the given
+// private key. The private key must match the PeerID field of the PeerRecord.
+func (r *PeerRecord) Sign(privKey crypto.PrivKey) (*record.Envelope, error) {
```

Comment:
Let's not have a `Sign` method per payload type. Proposal:

Add two new methods to the `Record` interface: `Domain() string` and `Codec() []byte`. The `record.MakeEnvelopeWithRecord` could then turn into the `record.Seal(crypto.PrivKey, record.Record)` method we talked about above, and you can remove the redundant parameters, as they can be obtained from the record instance.